jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/upload/EditLambdaDialog.kt

```diff
@@ -0,0 +1,363 @@
+            }
+
+            future.whenComplete { function, error ->
+                when {
+                        function != null -> {
+                            notifyInfo(
+                                message("lambda.function_created.notification", functionDetails.name),
+                                project = project
+                            )
+                        }
+                        error is Exception -> error.notifyError(title = "")
+                    }
+                }
+            close(OK_EXIT_CODE)
+        }
+    }
+
+    private fun updateLambda() {
+        if (updateAction.isEnabled) {
+            updateAction.isEnabled = false
+
+            val functionDetails = viewToFunctionDetails()
+            val lambdaClient: LambdaClient = AwsClientManager.getInstance(project).getClient()
+
+            ApplicationManager.getApplication().executeOnPooledThread {
+                LambdaFunctionCreator(lambdaClient).update(project, functionDetails)
+                    .thenAccept { runInEdt(ModalityState.any()) { close(OK_EXIT_CODE) } }
+                    .exceptionally { e ->
+                        runInEdt(ModalityState.any()) {
+                            setErrorText(e.message)
+                            updateAction.isEnabled = true
+                        }
+                        null
+                    }
+            }
+        }
+    }
+
+    private fun viewToFunctionDetails(): FunctionUploadDetails = FunctionUploadDetails(
+        name = view.name.text!!,
+        handler = view.handler.text,
+        iamRole = view.iamRole.selected()!!,
+        runtime = view.runtime.selected()!!,
+        description = view.description.text,
+        envVars = view.envVars.envVars,
+        timeout = view.timeout.text.toInt()
+    )
+
+    private inner class DeployLambdaAction : OkAction() {
+        init {
+            putValue(Action.NAME, message("lambda.upload.deploy_button.title"))
+        }
+
+        override fun doAction(e: ActionEvent?) {
+            // We normally don't validate the deploy settings in case they are editing settings only, but they requested
+            // to deploy so start validating that too
+            validateDeploySettings = true
+            super.doAction(e)
+            deployLambda()
+        }
+    }
+
+    // Using an OkAction to force the validation logic to trigger as well
+    private inner class UpdateLambdaSettingsAction : OkAction() {
+        init {
+            putValue(Action.NAME, message("lambda.upload.update_settings_button.title"))
+            putValue(DEFAULT_ACTION, null)
+        }
+
+        override fun doAction(e: ActionEvent?) {
+            super.doAction(e)
+            updateLambda()
+        }
+    }
+
+    private companion object {
+        const val LAMBDA_PRINCIPAL = "lambda.amazonaws.com"
+
+        @Language("JSON")
+        val DEFAULT_ASSUME_ROLE_POLICY = """
```

Comment:
These policies probably should not live in the dialog source file.

Reply:
Moved entire companion out to `LambdaPolicies.kt`